fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/sink/FlinkSink.java

```diff
@@ -102,24 +121,42 @@ static class UpsertSinkWriterBuilder implements SinkWriterBuilder<UpsertSinkWrit
         private final RowType tableRowType;
         private final @Nullable int[] targetColumnIndexes;
         private final boolean ignoreDelete;
+        private final int numBucket;
+        private final KeyEncoder bucketKeyEncoder;
+        private final boolean sinkReHash;
 
         UpsertSinkWriterBuilder(
                 TablePath tablePath,
                 Configuration flussConfig,
                 RowType tableRowType,
                 @Nullable int[] targetColumnIndexes,
-                boolean ignoreDelete) {
+                boolean ignoreDelete,
+                int numBucket,
+                KeyEncoder bucketKeyEncoder,
+                boolean sinkReHash) {
             this.tablePath = tablePath;
             this.flussConfig = flussConfig;
             this.tableRowType = tableRowType;
             this.targetColumnIndexes = targetColumnIndexes;
             this.ignoreDelete = ignoreDelete;
+            this.numBucket = numBucket;
+            this.bucketKeyEncoder = bucketKeyEncoder;
+            this.sinkReHash = sinkReHash;
         }
 
         @Override
         public UpsertSinkWriter createWriter() {
             return new UpsertSinkWriter(
                     tablePath, flussConfig, tableRowType, targetColumnIndexes, ignoreDelete);
         }
+
+        @Override
+        public DataStream<RowData> addPreWriteTopology(DataStream<RowData> input) {
+            return sinkReHash
+                    ? input.partitionCustom(
+                            (bucketId, numPartitions) -> bucketId % numPartitions,
+                            new RowDataKeySelector(bucketKeyEncoder, numBucket))
+                    : input;
```

Comment:
It would be better to customize the the shuffle name on the topology graph on Web UI. Currently, it is "CUSTOM", would be better to name "BUCKET_SHUFFLE" to be more meaningful for users. 

This might need to call more internal methods on DataStream.

Reply:
Besides, this doesn't handle the case of partitioned table? I mean, data will be skew for partitioned table if `task parallelism` and `bucket number` are  not divisible.

Reply:
Done!